### ISSUE TITLE
Improve matching quality

### ISSUE THREAD

**Issue by tarioch** [johannesjh/smart_importer#23]
I now imported my historical data from gnucash into beancount and I'm trying to match new imports with the existing data.
Right now the matches are really terrible.

I'd like to figure out why that's the case and improve the matching to actually be usable for me.

Points that I can think of
* imported narrations are "chatty" they contain some parts which are the same or almost the same in a lot of other transactions and only a small part is actually "interesting"
* suggestions are independent of amount => #22
* not sure if this is taken into account but I think the frequency of a certain transaction "type" should also be important. E.g. if I have a transaction "type" T1 which I have 100 times and a similar type T2 which occurs 1 time, it's way more likely that T1 is the correct match

**Comment by johannesjh** [comment]
Sorry to hear that. 

You may want to inspect the trained model, e.g., see https://eli5.readthedocs.io/en/latest/tutorials/sklearn-text.html

The points 1 and 3 that you mentioned above should already be taken into account by the machine learning:

- Chattiness: I don't think applying feature selection (to reduce vocabulary size) would improve predictions, but I am not sure.

- Frequencies: These are certainly taken into account already.

One idea would be to use a grid search for optimal parameters. Compare e.g. 

- http://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html
- http://scikit-learn.org/stable/auto_examples/plot_compare_reduction.html#sphx-glr-auto-examples-plot-compare-reduction-py

...but first you should probably inspect the trained pipeline.

**Comment by tarioch** [comment]
I already played around some with GridSearchCV and I was able to tune the parameters a little bit but it did not give a huge improvement. 
I started now to add the filtering for probability and that seems to help, if I only take the ones with > 60% probability I'm getting rid of the false positives.
Thanks for the link will certainly look some more into those.
As for 3, if I understood it correctly this gets taken care of if I add class_weight='balanced' to the svc and it seems to improve the situation for me.

**Comment by johannesjh** [comment]
oh right, looks like we should definitely add `class_weight='balanced'`

PS regarding `class_weight`:
I found the explanation from https://stackoverflow.com/a/30982811 nice and easy to understand

> For how `class_weight` works: It penalizes mistakes in samples of `class[i]` with `class_weight[i]` instead of 1. So higher class-weight means you want to put more emphasis on a class. From what you say it seems class 0 is 19 times more frequent than class 1. So you should increase the `class_weight` of class 1 relative to class 0, say {0:.1, 1:.9}.
If the `class_weight` doesn't sum to 1, it will basically change the regularization parameter.

**Comment by tarioch** [comment]
Have now done some improvements and raised some more bugs for individual problems, just found #27 which I think might be a major issue in my case, will get that one fixed and then check if the matching improves

**Comment by tarioch** [comment]
Ok, looks like after fixing #27 the matching is way, way, way better.
Now I think I have to retune the pipeline as it looks like I'm getting better results with master vs the branch.
But I would still like to have the probability in there for limiting the predictions as well as the suggestions.

**Comment by tarioch** [comment]
Some more findings:
- SVC with 'linear' seems to give the best matches without further tuning
- SVC with probability=True supports predict_proba method, but the results there don't (always) match what predict() returns

**Comment by tarioch** [comment]
Closing this issue for now, #27 was the major issue.
I'm not able to get the probability working without making the prediction worse.
